scalanet/src/io/iohk/scalanet/peergroup/CloseableQueue.scala

```diff
@@ -0,0 +1,95 @@
+package io.iohk.scalanet.peergroup
+
+import cats.implicits._
+import cats.effect.concurrent.{TryableDeferred, Deferred}
+import monix.catnap.ConcurrentQueue
+import monix.eval.Task
+import monix.execution.{BufferCapacity, ChannelType}
+import scala.util.{Left, Right}
+
+/** Wraps an underlying concurrent queue so that polling can return None when
+  * the producer side is finished, or vice versa the producer can tell when
+  * the consumer is no longer interested in receiving more values.
+  *
+  *
+  * @param closed indicates whether the producer side has finished and whether
+  * the messages already in the queue should or discarded (true) or consumed (false).
+  * @param queue is the underlying message queue
+  */
+class CloseableQueue[A](
+    closed: TryableDeferred[Task, Boolean],
+    queue: ConcurrentQueue[Task, A]
+) {
+  import CloseableQueue.Closed
+
+  /** Fetch the next item from the queue, or None if the production has finished
+    * and the queue has been emptied.
+    */
+  def next(): Task[Option[A]] =
+    closed.tryGet.flatMap {
+      case Some(true) =>
+        Task.pure(None)
+
+      case Some(false) =>
+        queue.tryPoll
+
+      case None =>
+        Task.race(closed.get, queue.poll).flatMap {
+          case Left(_) =>
+            next()
+          case Right(item) =>
+            Task.pure(Some(item))
+        }
+    }
+
+  /** Stop accepting items in the queue. Clear items if `discard` is true, otherwise let them be drained.
+    * If the queue is already closed it does nothing; this is because either the producer or the consumer
+    * could have closed the queue before.
+    */
+  def close(discard: Boolean): Task[Unit] =
+    closed.complete(discard).attempt >> queue.clear.whenA(discard)
+
+  /** Try to put a new item in the queue, unless the capactiy has been reached or the queue has been closed. */
+  def tryOffer(item: A): Task[Either[Closed, Boolean]] =
```

Comment:
maybe it is matter of opinion, why not returning failed task when some one tries to put item in closed queue ?

Reply:
Because it's not an exceptional condition, it's better in this case on a "closeable queue" to force people to think about the possibility that the offering will not be taken. I wouldn't want the producer to bomb just because the consumer isn't interested in the data any more and they forgot to catch the exception: they should check why the offering failed (closed or throttled) or ignore it because there's nothing they can do about it anyway. Similarly the underlying `tryOffer` doesn't throw an exception, it returns a boolean that you may or may not check.

My rule of thumb is that if the method is expected to fail, such as a `try` then or a validation function then it's good to use an `Either` to represent the types (e.g. I wouldn't know what exception to look for just by looking at the type), and I make the error types inherit `NoStackTrace` so the caller can easily call `.rethrow` on the thing if they rather want just exceptions than handle it the cumbersome way.

Reply:
oh that makes sense. i.e the whole point of having `closeable queue` is to make closing representable errror condition.